File: assets/codebase/menu/DemoDiscReturnToMenuComponent.cs
namespace city.menu {
    /// <summary>
    /// Returns the active demo-disc scene to the main menu when the temporary back bind is pressed.
    /// </summary>
    public sealed class DemoDiscReturnToMenuComponent : UpdateComponent {
        /// <summary>
        /// Interactable host used by pointer-enabled return buttons.
        /// </summary>
        InteractableComponent BoundInteractable;

        /// <summary>
        /// Tracks whether the active pointer press began inside the bound interactable.
        /// </summary>
        bool PointerPressStartedInside;

        /// <summary>
        /// Gets or sets whether keyboard return bindings may trigger the menu load.
        /// </summary>
        public bool AllowKeyboardReturn { get; set; } = true;

        /// <summary>
        /// Gets or sets whether gamepad return bindings may trigger the menu load.
        /// </summary>
        public bool AllowGamepadReturn { get; set; } = true;

        /// <summary>
        /// Gets or sets whether pointer interaction on a sibling interactable may trigger the menu load.
        /// </summary>
        public bool AllowPointerReturn { get; set; } = true;

        /// <summary>
        /// Tracks whether this component already requested a return transition during its current lifetime.
        /// </summary>
        bool SceneLoadWasRequested;

        /// <summary>
        /// Binds one sibling interactable when the component is attached to an authored clickable host.
        /// </summary>
        /// <param name="entity">Owning entity.</param>
        public override void ComponentAdded(Entity entity) {
            base.ComponentAdded(entity);
            if (AllowPointerReturn) {
                TryBindInteractable();
            }
        }

        /// <summary>
        /// Performs per-frame input polling for the demo-disc return bind.
        /// </summary>
        public override void Update() {
            if (AllowPointerReturn) {
                TryBindInteractable();
            }

            InputSystem inputSystem = Core.Instance.Input;
            bool wasReturnPressed = (AllowKeyboardReturn && WasKeyboardReturnPressed(inputSystem))
                || (AllowGamepadReturn && WasGamepadReturnPressed(inputSystem));

            if (wasReturnPressed) {
                LoadResolvedMainMenuScene();
            }
        }

        /// <summary>
        /// Releases any bound pointer interactable subscription before the component instance is deleted.
        /// </summary>
        public void Dispose() {
            UnbindInteractable();
        }

        /// <summary>
        /// Releases any bound pointer interactable subscription when the component detaches from its owner.
        /// </summary>
        /// <param name="entity">Owning entity.</param>
        public override void ComponentRemoved(Entity entity) {
            UnbindInteractable();
            base.ComponentRemoved(entity);
        }

        /// <summary>
        /// Returns whether the current frame pressed one of the desktop return keys.
        /// </summary>
        /// <param name="inputSystem">Input system supplying the current frame state.</param>
        /// <returns>True when one desktop return key pressed this frame.</returns>
        bool WasKeyboardReturnPressed(InputSystem inputSystem) {
#if DESKTOP_PLATFORM
            return inputSystem.WasKeyPressed(Keys.Escape) || inputSystem.WasKeyPressed(Keys.Back);
#else
            return false;
#endif
        }

        /// <summary>
        /// Returns whether the current frame pressed the configured standard platform return action.
        /// </summary>
        /// <param name="inputSystem">Input system supplying the current frame state.</param>
        /// <returns>True when one return button pressed this frame.</returns>
        bool WasGamepadReturnPressed(InputSystem inputSystem) {
            if (inputSystem == null) {
                throw new ArgumentNullException(nameof(inputSystem));
            }

            return DemoDiscReturnInputUtils.WasReturnPressed(inputSystem);
        }

        /// <summary>
        /// Binds one sibling interactable so pointer-enabled hosts can trigger return-to-menu clicks.
        /// </summary>
        void TryBindInteractable() {
            if (!AllowPointerReturn) {
                return;
            } else if (BoundInteractable != null || Parent == null || Parent.Components == null) {
                return;
            }

            for (int componentIndex = 0; componentIndex < Parent.Components.Count; componentIndex++) {
                if (Parent.Components[componentIndex] is InteractableComponent interactable) {
                    BoundInteractable = interactable;
                    BoundInteractable.CursorEvent += HandleCursorEvent;
                    return;
                }
            }
        }

        /// <summary>
        /// Releases the current pointer-interactable binding when the host no longer owns this component.
        /// </summary>
        void UnbindInteractable() {
            if (BoundInteractable == null) {
                return;
            }

            BoundInteractable.CursorEvent -= HandleCursorEvent;
            BoundInteractable = null;
            PointerPressStartedInside = false;
        }

        /// <summary>
        /// Handles pointer press and release events routed from one bound clickable host.
        /// </summary>
        /// <param name="relativePosition">Pointer position relative to the interactable.</param>
        /// <param name="delta">Pointer delta reported by the shared interaction router.</param>
        /// <param name="interaction">Current pointer interaction state.</param>
        void HandleCursorEvent(int2 relativePosition, int2 delta, PointerInteraction interaction) {
            if (!AllowPointerReturn) {
                return;
            }

            if (interaction == PointerInteraction.Press) {
                PointerPressStartedInside = true;
                return;
            }
            if (interaction == PointerInteraction.Release) {
                bool shouldReturnToMenu = PointerPressStartedInside;
                PointerPressStartedInside = false;
                if (shouldReturnToMenu) {
                    LoadResolvedMainMenuScene();
                }
                return;
            }
            if (interaction == PointerInteraction.Leave) {
                PointerPressStartedInside = false;
            }
        }

        /// <summary>
        /// Loads the logical main menu scene after resolving any active scene-map override.
        /// </summary>
        void LoadResolvedMainMenuScene() {
            if (SceneLoadWasRequested) {
                return;
            }

            string resolvedSceneId = DemoDiscMainMenuSceneResolver.ResolveRuntimeSceneId();
            SceneLoadWasRequested = true;
            Core.Instance.SceneManager.LoadScene(resolvedSceneId, SceneLoadMode.Single);
        }
    }
}

File: windows-build-20260602-stack-box-direct-real-bepu-fix8/cooked/code/gameplay/DemoDiscReturnToMenuComponent.cpp
#ifdef DrawText
#undef DrawText
#endif
#include "DemoDiscReturnToMenuComponent.hpp"
#include "runtime/native_exceptions.hpp"
#include "runtime/native_string.hpp"
#include "runtime/array.hpp"
#include "DemoDiscReturnToMenuComponent.hpp"
#include "runtime/native_list.hpp"
#include "system/action.hpp"
#include "Component.hpp"
#include "Core.hpp"
#include "CoreInitializationOptions.hpp"
#include "Entity.hpp"
#include "InputGamepadState.hpp"
#include "InputSystem.hpp"
#include "InteractableComponent.hpp"
#include "Keys.hpp"
#include "PointerInteraction.hpp"
#include "RuntimeSceneCatalog.hpp"
#include "RuntimeSceneCatalogEntry.hpp"
#include "SceneLoadMode.hpp"
#include "SceneManager.hpp"
#include "SceneMapComponent.hpp"
#include "StandardPlatformAction.hpp"
#include "StandardPlatformInput.hpp"
#include "UpdateComponent.hpp"
#include "runtime/array.hpp"
#include "runtime/native_cast.hpp"
#include "runtime/native_exceptions.hpp"
#include "runtime/native_list.hpp"
#include "runtime/native_string.hpp"

DemoDiscReturnToMenuComponent::DemoDiscReturnToMenuComponent() : AllowGamepadReturn(true), AllowKeyboardReturn(true), AllowPointerReturn(true), BoundInteractable(), PointerPressStartedInside(), SceneLoadWasRequested()
{
}

bool DemoDiscReturnToMenuComponent::get_AllowGamepadReturn()
{
return this->AllowGamepadReturn;
}

void DemoDiscReturnToMenuComponent::set_AllowGamepadReturn(bool value)
{
this->AllowGamepadReturn = value;
}

bool DemoDiscReturnToMenuComponent::get_AllowKeyboardReturn()
{
return this->AllowKeyboardReturn;
}

void DemoDiscReturnToMenuComponent::set_AllowKeyboardReturn(bool value)
{
this->AllowKeyboardReturn = value;
}

bool DemoDiscReturnToMenuComponent::get_AllowPointerReturn()
{
return this->AllowPointerReturn;
}

void DemoDiscReturnToMenuComponent::set_AllowPointerReturn(bool value)
{
this->AllowPointerReturn = value;
}

void DemoDiscReturnToMenuComponent::ComponentAdded(Entity* entity)
{
UpdateComponent::ComponentAdded(entity);
    if (this->AllowPointerReturn)
    {
this->TryBindInteractable();
    }
}

void DemoDiscReturnToMenuComponent::ComponentRemoved(Entity* entity)
{
this->UnbindInteractable();
UpdateComponent::ComponentRemoved(entity);
}

void DemoDiscReturnToMenuComponent::Dispose()
{
this->UnbindInteractable();
}

void DemoDiscReturnToMenuComponent::Update()
{
    if (this->AllowPointerReturn)
    {
this->TryBindInteractable();
    }
InputSystem *inputSystem = Core::get_Instance()->get_Input();
const bool wasReturnPressed = (this->AllowKeyboardReturn && this->WasKeyboardReturnPressed(inputSystem)) || (this->AllowGamepadReturn && this->WasGamepadReturnPressed(inputSystem));
    if (wasReturnPressed)
    {
this->LoadResolvedMainMenuScene();
    }
}

bool DemoDiscReturnToMenuComponent::CanLoadRuntimeScene(std::string sceneId)
{
    if (String::IsNullOrWhiteSpace(sceneId))
    {
throw ([&]() {
auto __ctor_arg_00000092 = "Scene id must be provided.";
auto __ctor_arg_00000093 = "sceneId";
return new ArgumentException(__ctor_arg_00000092, __ctor_arg_00000093);
})();
    }
    if (Core::get_Instance() == nullptr)
    {
throw new InvalidOperationException("A core instance must exist before loading runtime scenes.");
    }
    if (Core::get_Instance()->get_SceneManager() == nullptr)
    {
throw new InvalidOperationException("Core scene manager must be initialized before runtime scene loading can occur.");
    }
CoreInitializationOptions *initializationOptions = Core::get_Instance()->get_InitializationOptions();
RuntimeSceneCatalog *sceneCatalog = initializationOptions != nullptr ? initializationOptions->get_SceneCatalog() : nullptr;
    if (sceneCatalog == nullptr)
    {
return true;    }
Array<RuntimeSceneCatalogEntry*> *entries = sceneCatalog->get_Entries();
    if (entries == nullptr)
    {
return false;    }
for (int32_t entryIndex = 0; entryIndex < entries->get_Length(); entryIndex++) {
RuntimeSceneCatalogEntry *entry = (*entries)[entryIndex];
    if (entry == nullptr)
    {
continue;
    }
    if (String::Equals(entry->get_SceneId(), sceneId, StringComparison::OrdinalIgnoreCase))
    {
return true;    }
}
return false;}

void DemoDiscReturnToMenuComponent::HandleCursorEvent(int2 relativePosition, int2 delta, PointerInteraction interaction)
{
    if (!this->AllowPointerReturn)
    {
return;    }
    if (interaction == PointerInteraction::Press)
    {
this->PointerPressStartedInside = true;
return;    }
    if (interaction == PointerInteraction::Release)
    {
const bool shouldReturnToMenu = this->PointerPressStartedInside;
this->PointerPressStartedInside = false;
    if (shouldReturnToMenu)
    {
this->LoadResolvedMainMenuScene();
    }
return;    }
    if (interaction == PointerInteraction::Leave)
    {
this->PointerPressStartedInside = false;
    }
}

void DemoDiscReturnToMenuComponent::LoadResolvedMainMenuScene()
{
    if (this->SceneLoadWasRequested)
    {
return;    }
const std::string resolvedSceneId = SceneMapComponent::ResolveSceneId(MainMenuSceneId);
    if (!this->CanLoadRuntimeScene(resolvedSceneId))
    {
return;    }
this->SceneLoadWasRequested = true;
Core::get_Instance()->get_SceneManager()->LoadScene(resolvedSceneId, SceneLoadMode::Single);
}

void DemoDiscReturnToMenuComponent::TryBindInteractable()
{
    if (!this->AllowPointerReturn)
    {
return;    }
else {
    if (this->BoundInteractable != nullptr || this->get_Parent() == nullptr || this->get_Parent()->get_Components() == nullptr)
    {
return;    }
}
for (int32_t componentIndex = 0; componentIndex < this->get_Parent()->get_Components()->get_Count(); componentIndex++) {
    InteractableComponent* interactable = he_cpp_try_cast<InteractableComponent>((*this->get_Parent()->get_Components()).get_Item(componentIndex));
    if (interactable != nullptr)
    {
this->BoundInteractable = interactable;
this->BoundInteractable->CursorEvent += &DemoDiscReturnToMenuComponent::HandleCursorEvent;
return;    }
}
}

void DemoDiscReturnToMenuComponent::UnbindInteractable()
{
    if (this->BoundInteractable == nullptr)
    {
return;    }
this->BoundInteractable->CursorEvent -= &DemoDiscReturnToMenuComponent::HandleCursorEvent;
this->BoundInteractable = nullptr;
this->PointerPressStartedInside = false;
}

bool DemoDiscReturnToMenuComponent::WasGamepadButtonPressed(InputGamepadState currentState, InputGamepadState previousState, InputGamepadButton button)
{
return currentState.IsButtonDown(button) && !previousState.IsButtonDown(button);}

bool DemoDiscReturnToMenuComponent::WasGamepadReturnPressed(InputSystem* inputSystem)
{
    if (inputSystem == nullptr)
    {
throw new ArgumentNullException("inputSystem");
    }
    if (Core::get_Instance() == nullptr)
    {
throw new InvalidOperationException("A core instance must exist before querying the standard platform return action.");
    }
return Core::get_Instance()->get_StandardPlatformInput()->WasActionPressed(StandardPlatformAction::Return);}

bool DemoDiscReturnToMenuComponent::WasKeyboardReturnPressed(InputSystem* inputSystem)
{
return inputSystem->WasKeyPressed(Keys::Escape) || inputSystem->WasKeyPressed(Keys::Back);}

